### PR TITLE
Add ERC721BatchTransfer and couple improvements

### PR DESCRIPTION
**Pull request by channing-magiceden**
- Add `ERC721BatchTransfer` to enable 721 token airdrop
  - Mainnet `0x38F7ba911f7efc434D29D6E39c814E9d4De3FEef`
  - Polygon `0x38F7ba911f7efc434D29D6E39c814E9d4De3FEef`
- Add gas estimate for expensive operations
- Refactor scripts a little bit

### CODE REVIEW

File: contracts/utils/ERC721BatchTransfer.sol [hunk 1]
```diff
@@ -0,0 +1,159 @@
+//SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.4;
+
+import { IERC721 } from "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+
+/**
+ * @title ERC721 Batch Transfer
+ *
+ * @notice ETransfer ERC721 tokens in batches to a single wallet or multiple wallets. This supports ERC721M and ERC721CM contracts.
```
Comment (wolfy-nft) [inline contracts/utils/ERC721BatchTransfer.sol:10]: minor typo `ETransfer`

File: contracts/utils/ERC721BatchTransfer.sol [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+    event BatchTransferToMultiple(
+        address indexed contractAddress,
+        uint256 amount
+    );
+
+    constructor() {}
+
+    modifier noZero() {
+        if (msg.sender == address(0)) revert InvalidCaller();
```
Comment (wolfy-nft) [inline contracts/utils/ERC721BatchTransfer.sol:32]: what is the purpose of this? how could msg.sender ever be the null address?

File: contracts/utils/ERC721BatchTransfer.sol [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        address indexed to,
+        uint256 amount
+    );
+
+    event BatchTransferToMultiple(
+        address indexed contractAddress,
+        uint256 amount
+    );
+
+    constructor() {}
```
Comment (wolfy-nft) [inline contracts/utils/ERC721BatchTransfer.sol:29]: safe to remove this

File: contracts/utils/ERC721BatchTransfer.sol [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+     * @param erc721Contract the address of the nft contract
+     * @param to the address that will receive the nfts
+     * @param tokenIds the list of tokens that will be transferred
+     */
+    function batchTransferToSingleWallet(
+        IERC721 erc721Contract,
+        address to,
+        uint256[] calldata tokenIds
+    ) external noZero {
+        uint256 length = tokenIds.length;
```
Comment (wolfy-nft) [inline contracts/utils/ERC721BatchTransfer.sol:48]: if tokenIds is an empty array, we are still emitting an BatchTransferToSingle event. will this ever be a problem? we might want to revert if its an empty array

File: contracts/utils/ERC721BatchTransfer.sol [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+     * @param erc721Contract the address of the nft contract
+     * @param to the address that will receive the nfts
+     * @param tokenIds the list of tokens that will be transferred
+     */
+    function safeBatchTransferToSingleWallet(
+        IERC721 erc721Contract,
+        address to,
+        uint256[] calldata tokenIds
+    ) external noZero {
+        uint256 length = tokenIds.length;
```
Comment (wolfy-nft) [inline contracts/utils/ERC721BatchTransfer.sol:74]: same case here about zero length array

File: contracts/utils/ERC721BatchTransfer.sol [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    /** 
+     * @notice Transfer multiple tokens to the same wallet using the ERC721.transferFrom method.
+     * @notice If you don't know what that means, use the `safeBatchTransferToSingleWallet` method instead
+     * @param erc721Contract the address of the nft contract
+     * @param to the address that will receive the nfts
+     * @param tokenIds the list of tokens that will be transferred
+     */
+    function batchTransferToSingleWallet(
+        IERC721 erc721Contract,
+        address to,
```
Comment (wolfy-nft) [inline contracts/utils/ERC721BatchTransfer.sol:45]: should we revert if `to` is a null address?

File: contracts/utils/ERC721BatchTransfer.sol [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+    /**
+     * @notice transfer multiple tokens to the same wallet using the `ERC721.safeTransferFrom` method
+     * @param erc721Contract the address of the nft contract
+     * @param to the address that will receive the nfts
+     * @param tokenIds the list of tokens that will be transferred
+     */
+    function safeBatchTransferToSingleWallet(
+        IERC721 erc721Contract,
+        address to,
```
Comment (wolfy-nft) [inline contracts/utils/ERC721BatchTransfer.sol:71]: revert if zero address?

File: contracts/utils/ERC721BatchTransfer.sol (hunk 2 above)
Comment (channing-magiceden) [reply contracts/utils/ERC721BatchTransfer.sol:32]: It's not null address. Address(0) is valid address - we just don't know who has / will have the private key. Just a practice to not allow address(0) as it's known to be used as "null".

File: contracts/utils/ERC721BatchTransfer.sol (hunk 4 above)
Comment (channing-magiceden) [reply contracts/utils/ERC721BatchTransfer.sol:48]: yeah, I don't think it's a problem. We emit event whenever the function is called. Anyway we don't have use case to listen to events.

File: contracts/utils/ERC721BatchTransfer.sol (hunk 5 above)
Comment (channing-magiceden) [reply contracts/utils/ERC721BatchTransfer.sol:74]: same above.

File: contracts/utils/ERC721BatchTransfer.sol (hunk 6 above)
Comment (channing-magiceden) [reply contracts/utils/ERC721BatchTransfer.sol:45]: You mean `address(0)`? We should allow it so we can batch burn.

File: contracts/utils/ERC721BatchTransfer.sol (hunk 7 above)
Comment (channing-magiceden) [reply contracts/utils/ERC721BatchTransfer.sol:71]: Same

File: contracts/utils/ERC721BatchTransfer.sol (hunk 2 above)
Comment (wolfy-nft) [reply contracts/utils/ERC721BatchTransfer.sol:32]: it is a valid address but nobody has the compute power to generate the private key for address(0) which is why it used as the burn address

it currently holds > $200M in assets
https://etherscan.io/address/0x0000000000000000000000000000000000000000

if we are using this internally then it doesnt matter, just calling out that its not a necessary check to have

Comment (channing-magiceden) [reply contracts/utils/ERC721BatchTransfer.sol]: Ok just remove the check to save more gas.

**Review by wolfy-nft — APPROVED** [review]

**Pull request merged**